amethyst_renderer/src/pipe/effect/mod.rs

```diff
@@ -70,91 +73,101 @@ impl<'a> ProgramSource<'a> {
 }
 
 #[derive(Derivative)]
-#[derivative(Clone, Debug, Eq, Hash, PartialEq)]
+#[derivative(Clone, Debug, Eq, PartialEq)]
 pub struct Effect {
     pub pso: PipelineState<Meta>,
-    #[derivative(Hash = "ignore")]
-    pub pso_data: Data,
-    #[derivative(Hash = "ignore")]
-    pub samplers: HashMap<String, Sampler>,
-    #[derivative(Hash = "ignore")]
-    pub const_bufs: HashMap<String, RawBuffer>,
+    pub data: Data,
+    const_bufs: HashMap<String, usize>,
+    globals: HashMap<String, usize>,
 }
 
 impl Effect {
-    pub fn new_simple_prog<'a, S>(vs: S, ps: S) -> EffectBuilder<'a>
-        where S: Into<&'a [u8]>
-    {
-        EffectBuilder::new_simple_prog(vs, ps)
+    pub fn update_global<N: AsRef<str>, T: ToUniform>(&mut self, name: N, data: T) {
+        if let Some(i) = self.globals.get(name.as_ref()) {
+            self.data.globals[*i] = data.convert();
+        }
     }
 
-    pub fn new_geom_prog<'a, S>(vs: S, gs: S, ps: S) -> EffectBuilder<'a>
-        where S: Into<&'a [u8]>
+    /// FIXME: Update raw buffer without transmute, use `Result` somehow.
+    pub fn update_buffer<N, T>(&self, name: N, data: &[T], enc: &mut Encoder)
+        where N: AsRef<str>, T: Pod
     {
-        EffectBuilder::new_geom_prog(vs, gs, ps)
+        if let Some(i) = self.const_bufs.get(name.as_ref()) {
+            let raw = &self.data.const_bufs[*i];
+            enc.update_buffer::<T>(unsafe { mem::transmute(raw) }, &data[..], 0);
+        }
     }
 
-    pub fn new_tess_prog<'a, S>(vs: S, hs: S, ds: S, ps: S) -> EffectBuilder<'a>
-        where S: Into<&'a [u8]>
+    /// FIXME: Update raw buffer without transmute.
+    pub fn update_constant_buffer<N, T>(&self, name: N, data: &T, enc: &mut Encoder)
+        where N: AsRef<str>, T: Copy
     {
-        EffectBuilder::new_tess_prog(vs, hs, ds, ps)
+        if let Some(i) = self.const_bufs.get(name.as_ref()) {
+            let raw = &self.data.const_bufs[*i];
+            enc.update_constant_buffer::<T>(unsafe { mem::transmute(raw) }, &data);
+        }
+    }
+
+    /// FIXME: Add support for arbitrary materials and textures.
+    pub fn draw(&self, model: &Model, enc: &mut Encoder) {
```

Comment:
@omni-viral Unfortunately, the solution is TBD. Having a single `Material` struct is nice and simple but ultimately is starting to look rather limiting. In an ideal scenario, users should be able to define their own materials with data, and passes compatible with one or more of said materials can draw them.

Maybe we could make `Material` a trait and associate each one with a compatible `Effect` variant? But that might prevent users from defining new material types at runtime, though.